packages/taquito-michelson-encoder/src/schema/event-schema.ts

```diff
@@ -0,0 +1,65 @@
+import {
+  MichelsonV1Expression,
+  MichelsonV1ExpressionExtended,
+  ScriptResponse,
+} from '@taquito/rpc';
+import { deepEqual } from './storage';
+
+export class EventSchema {
+  constructor(
+    public readonly tag?: string,
+    public readonly type?: MichelsonV1ExpressionExtended
+  ) {}
+
+  static fromMichelineValue(val: MichelsonV1ExpressionExtended) {
+    return new EventSchema(
+      val.annots?.[0],
+      val.args?.[0] as MichelsonV1ExpressionExtended
+    );
+  }
+
+  static fromRPCResponse(val: { script: ScriptResponse }): EventSchema[] {
+    const allEventSchema: EventSchema[] = [];
+    val.script.code.forEach((code) => {
+      if (
+        !('prim' in code) ||
+        code.prim !== 'code' ||
+        !('args' in code) ||
+        !code.args
+      ) {
+        return;
+      }
+      allEventSchema.push(...EventSchema.extractEventsRecursively(code.args));
+    });
+    return EventSchema.removeDuplicates(allEventSchema);
+  }
+
+  static removeDuplicates(events: EventSchema[]): EventSchema[] {
```

Comment:
since we're removing dupes, do you think there's any merit in also logging how many occurrences of each event?

Reply:
I don't know. If you prefer, I can do that.

Reply:
I think it's fine as is for now. We don't know if that would be a worth while feature yet.